[16.0][IMP] project_rating_nps : Change computation of fields based on rating_data methods

Honeyxilia opened the pull request:
- In the `rating/models/rating_data.py` file are functions which allow the `rating` model to obtain several values based on a rating score.
  However, these functions use an `assert` instruction which bound rating scores between 0 to 5, which doesn't fit our NPS system using scores ranging from 0 to 10.
  As such, we overload their compute method to either assign the value found by our NPS method, or a dummy value for unused fields.

- We can use this opportunity to render the NPS into a `rating.parent.mixin` field, since we use the rating count to compute the NPS

fcayre commented on project_rating_nps/tests/test_project.py:
```diff
@@ -36,3 +36,9 @@ def test_nps(self):
+        "Checks overload of _compute_rating_satisfaction_percentage method"
+        self.assertEqual(self.project.rating_percentage_satisfaction, -1)
+        self.assertEqual(self.project.rating_avg, 0)
+        self.assertEqual(self.project.rating_avg_percentage, 0)
```
Je me demande si on ne devrait pas plutôt simplement appeler `_compute_rating_satisfaction_percentage` pour vérifier qu'elle ne plante pas, mais vérifier des valeurs qui n'ont pas pas de sens me paraît... ne pas avoir de sens :) !

Honeyxilia replied:
Accéder aux attributs directement permet de vérifier que la méthode compute assigne bien les valeurs - sinon, il devrait y avoir une exception CacheMiss si l'attribut est appelé par une autre vue

fcayre replied:
OK alors tu peux remplacer self.assertEqual par self.assertNotEqual(False, ...) comme ça tu es sûre qu'une valeur est mise, mais tu te fous de laquelle.

Honeyxilia replied:
J'applique une valeur de 0 à deux de ces valeurs, qui sont en l'état égales à `False` - je peux appliquer toute autre valeur cependant

fcayre commented on project_rating_nps/models/__init__.py:
```diff
@@ -1,2 +1,2 @@
-from . import project
+from . import rating_parent_mixin
```
On ne peut pas faire ça dans un module nommé **project**_rating_nps car alors on ne s'attend pas à modifier le modèle de notation d'un autre modèle. Si tu veux aller dans cette direction, je pense qu'il est possible de créer un nouveau mixin, nommé par exemple rating_npm_parent_mixin, renommer ce module rating_nps, et créer un autre module project_rating_nps qui ferait hériter project de ce nouveau mixin. Mais c'est pas mal de boulot en plus, je pense qu'il vaut mieux faire un ticket pour après la migration.

Honeyxilia replied:
Okie, dans ce cas là, on peut réduire le scope de la PR, et de rajouter une simple surcharge à `rating.parent.mixin` dans le fichier `project_rating_nps/models/rating.py`

Je le fais donc ?

Honeyxilia committed:
[FIX] Overload compute method to avoid assert function

The _compute_rating_satisfaction_percentage method uses functions from the rating/models/rating_data.py file,
which contains an assert instruction restricting the rating scores between 0 and 5, conflicting with our NPS system
(ranging from 0 to 10).

As such, we entirely overload the compute method in order to assign values to the following fields :
- rating_count
  (since this field is used in the project.project kanban view to display the NPS rating, we assign it)
- rating_satisfaction_percentage (unused)
- rating_avg (unused)
- rating_avg_percentage(unused)

fcayre commented on project_rating_nps/models/rating.py:
```diff
@@ -159,3 +159,39 @@ def rating_apply(
+class RatingParentMixin(models.AbstractModel):
+    _inherit = "rating.parent.mixin"
+
+    def _compute_rating_percentage_satisfaction(self):
```
Il aurait été mieux de ne surcharger cette méthode que pour les projets, afin de ne pas modifier le comportement des autres modèles qui hériteraient de ce mixin. Il me semble que j'avais essayé en live et pas réussi, mais ça mérite de creuser pour y parvenir. Idéalement on ajouterait même un test avec une classe Model de test qui hérite de ce mixin et qui vérifie que le comportement original est préservé pour cette classe.

Honeyxilia replied:
je suis en train de rédiger des tests avec des classes de test c:

Honeyxilia committed:
[IMP] Assigned NPS compute method to rating_count

Since the total amount of ratings is already computed in the _compute_net_promoter_score method,
we can simply assign rating_count in it, to avoid it being computed twice.

Honeyxilia committed:
[REF] Modified rating search domain to avoid searching tasks.

Since ratings for a resourcecan be looked for using a resource's parent - ie. tasks ratings through their project -,
we can directly look for ratings whose parent resource is the current project.

This avoids a project.task search.

Honeyxilia committed:
[REF] Replaced search_count operations by a single search operation

Rather than doing several searches on the database, we can fetch all of the ratings concerning
the current project in a single recordset, then filtering on the rating value to compute the amount
of total ratings, detractors and promoters.

fcayre commented on project_rating_nps/models/project.py:
```diff
@@ -16,17 +16,12 @@ class Project(models.Model):
-                ("res_model", "=", task_ids._name),
-                ("res_id", "in", task_ids.ids),
+                ("parent_res_model", "=", record._name),
+                ("parent_res_id", "in", record.ids),
```
Why not `("parent_res_id", "=", record.id),` ?

Honeyxilia replied:
to reduce the amount of changes in the commit - we can change it to simplify the domain search c:

fcayre commented on project_rating_nps/models/project.py:
```diff
@@ -31,19 +31,15 @@ def _compute_net_promoter_score(self):
+                detractors_count = len(ratings.filtered_domain([("rating", "<=", 6)]))
                 record.net_promoter_score = int(
-                    100 * (1.0 * promoters_count - detractors_count) / total_count
+                    100 * (1.0 * promoters_count - detractors_count) / len(ratings)
```
`record.rating_count` instead of `len(ratings)` would prevent length recomputation.

fcayre commented on project_rating_nps/models/project.py:
```diff
@@ -34,6 +37,7 @@ def _compute_net_promoter_score(self):
                 ),
             ]
             total_count = self.env["rating.rating"].search_count(base_domain)
```
I cannot comment on an unmodified line, but I think we do not need to convert datetime to string anymore in the base_domain.

Honeyxilia replied:
indeed - I can remove it in a dedicated commit to simplify further the domain search.

Tangentially, the `_compute_rating_satisfaction_percentage` method for fetching the ratings uses `write_date` as opposed to `create_date` for the ratings - shall we change it too ?

Honeyxilia committed:
[IMP] Removed datetime to string convertion in search domain

We can directly use datetimes in the search domain - as such, we remove the datetime to string conversion
in order to increase lisibility of the search domain.

fcayre commented on project_rating_nps/models/project.py:
```diff
@@ -23,13 +23,7 @@ def _compute_net_promoter_score(self):
-                        fields.datetime.now() - timedelta(days=30)
-                    ),
-                ),
+                ("create_date", ">=", fields.datetime.now() - timedelta(days=30)),
```
Tu me dis que le code de Odoo SA utilise write_date au lieu de create_date et en effet ça paraît judicieux ici, +1 pour faire la modif et fusionner dans la foulée car le tout me paraît nickel :+1:

Honeyxilia replied:
:star: je fais un commit dédié afin d'expliquer le raisonnement du changement

Honeyxilia committed:
[IMP] Replaced create_date search argument by write_date

In the rating module, the ratings are searched relative to their write_date and not their create_date.
This allows for ratings filled in late to still count towards the satisfaction percentage.

We wish to apply the same reasonning to the NPS system, so we use write_date now.

Honeyxilia commented:
/ocabot rebase

Honeyxilia commented:
/ocabot merge patch
